elytron/WFCORE-3596-named-permission-sets.adoc

```diff
@@ -0,0 +1,91 @@
+:email:             fjuma@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+Currently, permission mappings and constant permission mappers in the Elytron subsystem configuration
+reference individual permissions. This task is to introduce named permission sets, i.e., instead of
+having a permission mapping or constant permission mapper reference permissions, they would instead
+reference named permission sets.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3596[WFCORE-3596]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-978[EAP7-978]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* It should be possible to configure named permission sets in the Elytron subsystem configuration. Each permission
+  set should contain zero or more permissions. It should be possible to reference permission sets from permission
+  mappings and from constant permission mappers. As an example, the default configuration would be as shown below.
+  Notice that the default permissions will be in a default permission set, "default-permissions", that is referenced by
+  the "default-permission-mapper". For the provisioning tool, the default permission set will be the single place to add
+  permissions as subsystems are being added and it will be the single place to remove permissions as subsystems are being
+  removed.
+
+[source,xml]
+----
+<mappers>
+...
+    <simple-permission-mapper name="default-permission-mapper" mapping-mode="first">
+        <permission-mapping>
+            <principal name="anonymous"/>
+            <permission-set name="default-permissions"/>
+        </permission-mapping>
+        <permission-mapping match-all="true">
+            <permission-set name="login-permission" />
+            <permission-set name="default-permissions"/>
+        </permission-mapping>
+    </simple-permission-mapper>
+...
+</mappers>
+
+<permission-sets>
+    <permission-set name="login-permission">
+        <permission class-name="org.wildfly.security.auth.permission.LoginPermission"/>
+    </permission-set>
+    <permission-set name="default-permissions">
+        <permission class-name="org.wildfly.extension.batch.jberet.deployment.BatchPermission" module="org.wildfly.extension.batch.jberet" target-name="*"/>
+        <permission class-name="org.wildfly.transaction.client.RemoteTransactionPermission" module="org.wildfly.transaction.client"/>
+        <permission class-name="org.jboss.ejb.client.RemoteEJBPermission" module="org.jboss.ejb-client"/>
+    </permission-set>
+</permission-sets>
+----
+
+* When transforming the model for legacy slaves, the permission sets should be in-lined.
```

Comment:
Could you elaborate more on this? Does that mean <permission-set> will be resolved to list of <permission> ?

Reply:
Is there also automatic transformation to new version of model. I mean when I take WF12 standalone.xml and will put it in WF13 and restart the server I will end up with permission set representation?

In other words when I use WF12 standalone.xml in WF13, will I be able to manipulate permission-sets in permission-mapper ?

Reply:
When transforming the model, the permission-set will be replaced with the permissions that make up the permission-set. 

XML configuration for prior versions will be converted to the new version, i.e., each list of permissions will be converted to a permission set.